### ISSUE TITLE
stub-resolv.conf does not contains 127.0.0.53, if DNSStubListener=no but DNSStubListenerExtra=0.0.0.0.

### ISSUE THREAD

**Issue by yw662** [systemd/systemd#17773]
**systemd version the issue has been seen with**
 > …
systemd 247 (247.50-1-arch)
+PAM +AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +ZSTD +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid

**Used distribution**
 > …
archlinux, inside container

**Linux kernel version used** (`uname -a`)
 > …
Linux penguin 5.4.58-07649-ge120df5deade #1 SMP PREEMPT Wed Aug 26 04:56:33 PDT 2020 x86_64 GNU/Linux

**CPU architecture issue was seen on**
 > …
x86_64, GenuineIntel

**Expected behaviour you didn't see**
 > …
Even if DNSStubListener is unset, since DNSStubListenerExtra covers localhost, resolved should still list itself in stub-resolv.conf (then /etc/resolv.conf through symlink).

**Unexpected behaviour you saw**
 > …
127.0.0.53 (as well as 127.0.0.1) is not in stub-resolv.conf. Instead, all configured upstreams are listed.

**Steps to reproduce the problem**
 > …

use resolved.conf like this:
```
[Resolve]
DNSStubListener=no
DNSStubListenerExtra=0.0.0.0
DNS=1.1.1.1 8.8.8.8 9.9.9.9
```

then restart resolved, then check stub-resolv.conf (at /run/systemd/resolve/stub-resolv.conf). It will be like
```
nameserver 1.1.1.1
nameserver 8.8.8.8
nameserver 9.9.9.9
```

**Comment by yw662** [comment]
Here is a workaround, but not so good:

Let stubListener be tcp and stubListenerExtra be udp (or vise versa), like this:
```
[Resolve]
DNSStubListener=tcp
DNSStubListenerExtra=udp:0.0.0.0
```

**Comment by poettering** [comment]
You turned off the primary stub, hence we will use the upstream DNS server info.

The extra stubs are entirely independent of that, and are facing to other systems (for example, local containers or VMs or so). 

If you want resolved to answer local requests, then leave the local stub enabled.

**Comment by yw662** [comment]
Yes I understand that, but, since the extra stub may also cover localhost, it could also be used as local resolver.

Besides, is there a way to tell the extra stub to listen to "every addresses but 127.9.0.53" ? The existence of the local stub make it impossible for the extra stub to listen to 0.0.0.0, that is why it is an issue.

**Comment by andrevmatos** [comment]
You don't need to `DNSStubListener=tcp` (which actually breaks resolving on some tools which try to resolve on 127.0.0.53 but receive a response from 127.0.0.1), just specifying the udp protocol in the Extra listener allows resolved to listen on both `tcp|udp:127.0.0.53:53` (default stub) and `udp:0.0.0.0:53`:
```
[Resolve]
DNSStubListenerExtra=udp:0.0.0.0
```
